### PR TITLE
Rewrite six.b(...) to b-literal

### PR DESCRIPTION
**Pull request by mxr**
Taking a stab at this. Not sure if this is the best approach or if there are edge cases I missed. This is part of #59.

### CODE REVIEW

File: pyupgrade.py [hunk 1]
```diff
@@ -1197,6 +1207,9 @@ def _replace_attr(i, mapping, node):
                 func_args, end = _parse_call_args(tokens, i + 1)
                 template = SIX_CALLS[node.func.id]
                 _replace_call(tokens, i, end, func_args, template)
+        elif token.offset in visitor.literal_attrs:
+            func_args, end = _parse_call_args(tokens, i + 3)
+            _replace_call(tokens, i, end, func_args, 'b{args[0]}')
```
Comment (asottile) [inline pyupgrade.py]: if it's this easy (?) I think this can just be an entry in the `SIX_CALLS` mapping above

Comment (asottile) [reply pyupgrade.py]: oh except it has to be a literal 🤔

Comment (mxr) [reply pyupgrade.py]: It seems like using a different set of attrs is a cleaner way to implement this special case

**Comment by asottile** [comment]
looks like the only edge case (that isn't currently handled) is the literal has to be entirely ascii characters (even though `b` is implemented as `.encode('latin-1')`

so `b('\x12\xef')` is fine, but `b('£')` is not safely rewritable (despite being in latin-1)

**Comment by mxr** [comment]
Good point. I'll update this tomorrow

**Comment by mxr** [comment]
Maybe I'm not understanding what you mean by entirely ascii characters? 

```console
$ python3.7 -c "print('\x123\xef'.isascii())"
False
```

Also I realized I have to detect `from six import b`

**Comment by asottile** [comment]
> Maybe I'm not understanding what you mean by entirely ascii characters?
> 
> ```
> $ python3.7 -c "print('\x123\xef'.isascii())"
> False
> ```
> 
> Also I realized I have to detect `from six import b`

Not the string itself, but the source.

Hope this demonstrates:

```pycon
>>> from six import b
>>> b('foo') == b'foo'
True
>>> b('\xa1') == b'\xa1'
True
>>> b('£')
b'\xa3'
>>> b('£') == b'£'
  File "<stdin>", line 1
SyntaxError: bytes can only contain ASCII literal characters.
```

**Comment by mxr** [comment]
@asottile let me know how this looks?

File: pyupgrade.py [hunk 2]
```diff
@@ -669,6 +669,20 @@ def _is_bytestring(s):
...
+        return s.isascii()
+    else:  # pragma: no cover (<py37)
+        try:
+            s.encode('ascii')
+        except UnicodeDecodeError:  # pragma: no cover (py2)
+            return False
+        except UnicodeEncodeError:  # pragma: no cover (py2+)
+            return False
+        else:
+            return True
```
Comment (asottile) [inline pyupgrade.py]: some of these branches aren't necessary -- the tokens you get will always be text -- did this get copy pasted from somewhere?

File: pyupgrade.py [hunk 3]
```diff
@@ -1086,6 +1101,22 @@ def visit_Call(self, node):
...
+                isinstance(node.args[0], ast.Str)
+        ):
+            self.literal_attrs[_ast_to_offset(node)] = node
+        elif (
+                isinstance(node.func, ast.Name) and
+                node.func.id == 'b' and
+                node.func.id in self.six_from_imports and
+                isinstance(node.args[0], ast.Str)
+        ):
+            self.literal_attrs[_ast_to_offset(node)] = node
```
Comment (asottile) [inline pyupgrade.py]: any way that `is_six_attr` / `is_six_name` can be reused?

File: pyupgrade.py [hunk 4]
```diff
@@ -1197,6 +1228,20 @@ def _replace_attr(i, mapping, node):
...
+            template = 'b{args[0]}'
+            if tokens[i + 1].src == '(' and _is_ascii(tokens[i + 2].src):
+                func_args, end = _parse_call_args(tokens, i + 1)
+                _replace_call(tokens, i, end, func_args, template)
+            elif (
+                    tokens[i + 1].src == '.' and
+                    tokens[i + 2].src == node.func.attr and
+                    tokens[i + 3].src == '(' and
+                    _is_ascii(tokens[i + 4].src)
+            ):
```
Comment (asottile) [inline pyupgrade.py]: ugly, but I think this current breaks:

```
six.b(    "eurosymbol")
```

File: pyupgrade.py (hunk 2 above)
Comment (mxr) [reply pyupgrade.py]: I think I was playing with b’’ strings in the interpreter. Will revisit this

File: pyupgrade.py (hunk 4 above)
Comment (mxr) [reply pyupgrade.py]: I see. Do the other six fixers handle this?

Comment (asottile) [reply pyupgrade.py]: yes, this case differs because you have to inspect the arguments

File: pyupgrade.py (hunk 3 above)
Comment (mxr) [reply pyupgrade.py]: I was able to do it

**Pull request merged**